src/foam/nanos/approval/AuthenticatedApprovalDAOAuthorizer.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.approval',
+  name: 'AuthenticatedApprovalDAOAuthorizer',
+  implements: ['foam.nanos.auth.Authorizer'],
+
+  javaImports: [
+    'foam.core.X',
+    'foam.core.FObject',
+    'foam.dao.*',
+    'foam.mlang.order.Comparator',
+    'foam.mlang.predicate.Predicate',
+    'foam.nanos.auth.AuthService',
+    'foam.nanos.auth.AuthenticationException',
+    'foam.nanos.auth.AuthorizationException',
+    'foam.nanos.auth.User',
+    'foam.util.SafetyUtil',
+    'foam.nanos.approval.ApprovalStatus',
+    'foam.nanos.approval.ApprovalRequest',
+  ],
+
+  methods: [
+    {
+      name: 'authorizeOnCreate',
+      javaCode: `
+        AuthService authService = (AuthService) x.get("auth");
+        Long userId = ((User) x.get("user")).getId();
+        if ( ! authService.check(x, "approval.create") ) {
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'authorizeOnRead',
+      javaCode: `
+        User user = (User) x.get("user");
+        if ( user == null || ! SafetyUtil.equals(user.getId(), ((ApprovalRequest)obj).getApprover()) ) {
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'authorizeOnUpdate',
+      javaCode: `
+        ApprovalRequest approvalRequest = (ApprovalRequest) oldObj;
+        Long userId = ((User) x.get("user")).getId();
+        AuthService authService = (AuthService) x.get("auth");
+        if ( ! authService.check(x, "approval.update.*") && ! SafetyUtil.equals(approvalRequest.getApprover(), userId) ) {
```

Comment:
Shouldnt this be approval.update. + oldObj.getId(), since its not a global check. Automatically a user with approval.update.* will be able to update all approvals.

Might make more sense if authorizeOnRead, authorizeOnUpdate and authorizeOnDelete check if the user id is equal to the approver id rather than an explicit permission string check because then its either check if a user has wildcard and they'll see all approval requests or check if they have a permission to individually read that request itself which I don't think works best with approvals. I remember this specific check for if the user is the approver was done in the approvalDAO somewhere before @jlhughes @lchanmann @noodlemoodle

Reply:
I see ruby talked about it in read

Reply:
makes sense to me for read and update, but i'm not sure if users should be able to delete approvalrequests unless they have explicit permission

Reply:
So what should be done for delete?

Reply:
Ya so delete should be then just admin or the system:

```
foam.nanos.auth.User user = (foam.nanos.auth.User) x.get("user");
if ( user != null && ( user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) return;
throw new AuthorizationException("Approval requests can only be created by the system");
```

Is what is on the LiquidApprovalRequestAuthorizer